compiler/src/iree/compiler/Codegen/ExternalInterfaces/Utils.cpp

```diff
@@ -7,19 +7,93 @@
 #include "iree/compiler/Codegen/ExternalInterfaces/Utils.h"
 
 #include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenInterfaces.h"
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenTypes.h"
 #include "iree/compiler/Codegen/Dialect/Codegen/Utils/Utils.h"
+#include "iree/compiler/Dialect/Encoding/IR/EncodingTypes.h"
 #include "llvm/Support/Casting.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
 #include "mlir/IR/Attributes.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypeInterfaces.h"
 #include "mlir/IR/MLIRContext.h"
 
 namespace mlir::iree_compiler::IREE {
 using Codegen::MaterializeEncodingInfo;
 
+Value calculateStorageSizeInBytesImpl(Attribute attr, Location loc,
+                                      OpBuilder &builder, RankedTensorType type,
+                                      ValueRange dynamicDims) {
+  auto deviceLayoutAttr = cast<IREE::Codegen::LayoutAttrInterface>(attr);
+  MaterializeEncodingInfo encodingInfo = deviceLayoutAttr.getEncodingInfo(type);
+  SmallVector<int64_t> paddedShape(type.getShape());
+  SmallVector<Value> paddedDynamicDims(dynamicDims.begin(), dynamicDims.end());
+  for (auto [dim, size] : llvm::zip_equal(encodingInfo.innerDimsPos,
+                                          encodingInfo.innerTileSizes)) {
+    // Only VMVX backend has dynamic inner tile sizes when ukernel is enabled.
+    // It assumes that the padding size is 16. Ideally, the logic should be
+    // moved to VMVX implementation details. However, we cook the logic here to
+    // reduce code duplication.
+    if (ShapedType::isDynamic(size)) {
```

Comment:
I'm slightly concerned that if and when we ever have a dynamic size outside of VMVX, this will silently cause a misallocation (if that future target requires a different size than 16). It would be nice to assert here that the target is in fact VMVX. It seems that that might require passing some additional parameter down to this function.

Reply:
Also, consider merging this `if (ShapedType::isDynamic(size))` with the `if (type.isDynamicDim(dim)) {` below.

Reply:
Good point, let me add an assertion. We don't need to pass an additional parameter down to the function, because we can check if the attribute can be cast to VMVX* attribute or not.

I don't follow how to merge two statement. One is the tile size and the other is the dimension size. The `size` can be used in both branch.

Reply:
Oh right, I had misread that code.